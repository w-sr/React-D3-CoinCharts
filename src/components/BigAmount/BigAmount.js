import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import currencyFormatter from "currency-formatter";

import { color, fontSize, fontWeight } from "../../styles/constants";
import { DEFAULT_PROPS, PROPTYPES } from "../../constants";

export const MINUS_SYMBOL = "\u2212";
export const PLUS_SYMBOL = "+";

const StyledBigAmount = styled.div`
  color: ${color.slateDark};
  font-size: ${fontSize.mega};
  font-weight: ${fontWeight.regular};
`;

const Superscript = styled.span.attrs({
  color: props => props.color || color.slateDark,
})`
  color: ${props => props.color};
  font-size: 30px;
  font-weight: ${fontWeight.medium};
  position: relative;
  top: -13px;
  vertical-align: baseline;
`;

const BigAmount = ({ type, showPlusMinusSymbol, value, currency }) => {
  switch (type) {
    case "currency": {
      const { decimalSeparator, spaceBetweenAmountAndSymbol, symbol, symbolOnLeft } = currencyFormatter.findCurrency(
        currency,
      );
      const formattedValue = currencyFormatter.format(Math.abs(value), {
        code: currency,
        symbol: "",
      });
      const [integerValue, decimalValue] = formattedValue.split(decimalSeparator);
      const optionalSpacing = spaceBetweenAmountAndSymbol ? " " : "";

      return (
        <StyledBigAmount>
          {showPlusMinusSymbol && value > 0 && <Superscript color={color.positive}>{PLUS_SYMBOL}</Superscript>}
          {showPlusMinusSymbol && value < 0 && <Superscript>{MINUS_SYMBOL}</Superscript>}
          {symbolOnLeft && <Superscript>{`${symbol}${optionalSpacing}`}</Superscript>}
          {integerValue && <span>{integerValue}</span>}
          {decimalValue && <Superscript>{`${decimalSeparator}${decimalValue}`}</Superscript>}
          {!symbolOnLeft && <Superscript>{`${optionalSpacing}${symbol}`}</Superscript>}
        </StyledBigAmount>
      );
    }
    case "percentage": {
      const formattedValue = Math.abs(value).toFixed(2);

      return (
        <StyledBigAmount>
          {showPlusMinusSymbol && value > 0 && <Superscript color={color.positive}>{PLUS_SYMBOL}</Superscript>}
          {showPlusMinusSymbol && value < 0 && <Superscript>{MINUS_SYMBOL}</Superscript>}
          <span>{formattedValue}</span>
          <Superscript>%</Superscript>
        </StyledBigAmount>
      );
    }
    default:
      return (
        <StyledBigAmount>
          <span>{value}</span>
        </StyledBigAmount>
      );
  }
};

BigAmount.propTypes = {
  value: PropTypes.number.isRequired,
  type: PROPTYPES.BIG_AMOUNT_TYPE,
  currency: PropTypes.string,
  showPlusMinusSymbol: PropTypes.bool,
};

BigAmount.defaultProps = {
  type: undefined,
  currency: DEFAULT_PROPS.CURRENCY,
  showPlusMinusSymbol: false,
};

export default BigAmount;
