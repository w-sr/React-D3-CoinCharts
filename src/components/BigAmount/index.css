.BigAmount .green {
  color: #61ca00;
}

.BigAmount .large-font {
  font-size: 48px;
  font-weight: 400;
  vertical-align: baseline;
}

.BigAmount .small-font {
  font-size: 30px;
  font-weight: 500;
  top: -13px;
  position: relative;
}
