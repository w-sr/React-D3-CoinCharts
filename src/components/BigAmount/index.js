import BigAmount from "./BigAmount";

export default BigAmount;
