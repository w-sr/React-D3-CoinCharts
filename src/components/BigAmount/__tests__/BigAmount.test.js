import React from "react";
import { render } from "react-testing-library";

import BigAmount from "../";
import { MINUS_SYMBOL, PLUS_SYMBOL } from "../BigAmount";

describe("<BigAmount />", () => {
  it("renders without crashing", () => {
    const props = {
      value: 99.99,
    };

    const { container } = render(<BigAmount {...props} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders without crashing (type: currency)", () => {
    const props = {
      value: 99.99,
      type: "currency",
    };

    const { container } = render(<BigAmount {...props} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
    expect(container.textContent).toEqual("$99.99");
  });

  it("renders without crashing (type: percentage)", () => {
    const props = {
      value: 99.99,
      type: "percentage",
    };

    const { container } = render(<BigAmount {...props} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
    expect(container.textContent).toEqual("99.99%");
  });

  it("renders PLUS_SYMBOL/MINUS_SYMBOL when `props.showPlusMinusSymbol` is true", () => {
    const props = {
      value: 99.99,
      type: "percentage",
      showPlusMinusSymbol: true,
    };

    const { container } = render(<BigAmount {...props} />);
    expect(container.textContent).toEqual(`${PLUS_SYMBOL}99.99%`);

    props.type = "currency";
    props.value = -99.99;
    render(<BigAmount {...props} />, { container });
    expect(container.textContent).toEqual(`${MINUS_SYMBOL}$99.99`);
  });

  it("not render PLUS_SYMBOL/MINUS_SYMBOL when `props.showPlusMinusSymbol` is false", () => {
    const props = {
      value: 99.99,
      type: "percentage",
      showPlusMinusSymbol: false,
    };

    const { container } = render(<BigAmount {...props} />);
    expect(container.textContent).toEqual("99.99%");

    props.type = "currency";
    props.value = -99.99;
    render(<BigAmount {...props} />, { container });
    expect(container.textContent).toEqual("$99.99");
  });
});
