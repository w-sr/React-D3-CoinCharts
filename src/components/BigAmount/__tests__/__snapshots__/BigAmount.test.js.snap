// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BigAmount /> renders without crashing (type: currency) 1`] = `
<div
  class="sc-bdVaJa jlxkRb"
>
  <span
    color="#4E5C6E"
    class="sc-bwzfXH fbCOIX"
  >
    $
  </span>
  <span>
    99
  </span>
  <span
    color="#4E5C6E"
    class="sc-bwzfXH fbCOIX"
  >
    .99
  </span>
</div>
`;

exports[`<BigAmount /> renders without crashing (type: percentage) 1`] = `
<div
  class="sc-bdVaJa jlxkRb"
>
  <span>
    99.99
  </span>
  <span
    color="#4E5C6E"
    class="sc-bwzfXH fbCOIX"
  >
    %
  </span>
</div>
`;

exports[`<BigAmount /> renders without crashing 1`] = `
<div
  class="sc-bdVaJa jlxkRb"
>
  <span>
    99.99
  </span>
</div>
`;
